### PR TITLE
ENH: add Makefile information helper

### PR DESCRIPTION
**Pull request by klauer**
## Description
* Start of a tool to help inspect `Makefile` information
* Currently allows for inspection of all makefile variables and the default make target
* Splits out simple `parse` to `whatrecord.parse` in order to make it more accessible. External imports relying on `.bin.parse.parse` should remain functional.

This method actually reveals a ton of information about the build when used with EPICS makefiles. It's a pretty interesting exercise, and I think we may be able to leverage this for more than I initially thought:

<details>

<summary>Variables</summary>

```
    "EPICSDEV": "/reg/g/pcds/epics-dev",
    "EPICSPROD": "/reg/g/pcds/epics",
    "EPICS_BASE": "/cds/group/pcds/epics/base/R7.0.3.1-2.0",
    "EPICS_BASE_HOST_BIN": "/cds/group/pcds/epics/base/R7.0.3.1-2.0/bin/linux-x86_64",
    "EPICS_BASE_HOST_LIB": "/cds/group/pcds/epics/base/R7.0.3.1-2.0/lib/linux-x86_64",
    "EPICS_BASE_HOST_LIBS": "ca Com",
    "EPICS_BASE_IOC_LIBS": "dbRecStd dbCore ca Com",
    "EPICS_BASE_PVA_CORE_LIBS": "pvAccessIOC pvAccessCA pvAccess pvData",
    "EPICS_BASE_VERSION": "7.0.5.0",
    "EPICS_CA_AUTO_ADDR_LIST": "YES",
    "EPICS_CA_BEACON_PERIOD": "15.0",
    "EPICS_CA_CONN_TMO": "30.0",
    "EPICS_CA_DEVELOPMENT_FLAG": "0",
    "EPICS_CA_MAINTENANCE_VERSION": "5",
    "EPICS_CA_MAJOR_VERSION": "4",
    "EPICS_CA_MAX_ARRAY_BYTES": "40000000",
    "EPICS_CA_MAX_SEARCH_PERIOD": "300",
    "EPICS_CA_MINOR_VERSION": "13",
    "EPICS_CA_REPEATER_PORT": "5065",
    "EPICS_CA_SERVER_PORT": "5064",
    "EPICS_DATABASE": "/cds/group/pcds/epics/base/R7.0.3.1-2.0",
    "EPICS_DATABASE_DEVELOPMENT_FLAG": "0",
    "EPICS_DATABASE_HOST_BIN": "/cds/group/pcds/epics/base/R7.0.3.1-2.0/bin/linux-x86_64",
    "EPICS_DATABASE_MAINTENANCE_VERSION": "5",
    "EPICS_DATABASE_MAJOR_VERSION": "3",
    "EPICS_DATABASE_MINOR_VERSION": "17",
    "EPICS_DEV_SNAPSHOT": "",
    "EPICS_EXTENSIONS": "/reg/g/pcds/epics/extensions/R0.2.0",
    "EPICS_HOST_ARCH": "linux-x86_64",
    "EPICS_LIBCOM_DEVELOPMENT_FLAG": "0",
    "EPICS_LIBCOM_MAINTENANCE_VERSION": "6",
    "EPICS_LIBCOM_MAJOR_VERSION": "3",
    "EPICS_LIBCOM_MINOR_VERSION": "17",
    "EPICS_MODIFICATION": "3",
    "EPICS_MODULES": "/cds/group/pcds/epics/R7.0.3.1-2.0/modules",
    "EPICS_NTYPES_DEVELOPMENT_FLAG": "0",
    "EPICS_NTYPES_MAINTENANCE_VERSION": "1",
    "EPICS_NTYPES_MAJOR_VERSION": "6",
    "EPICS_NTYPES_MINOR_VERSION": "0",
    "EPICS_PATCH_LEVEL": "1",
    "EPICS_PATCH_VSTRING": ".1",
    "EPICS_PVACLIENT_DEVELOPMENT_FLAG": "0",
    "EPICS_PVACLIENT_MAINTENANCE_VERSION": "0",
    "EPICS_PVACLIENT_MAJOR_VERSION": "4",
    "EPICS_PVACLIENT_MINOR_VERSION": "7",
    "EPICS_PVA_AUTO_ADDR_LIST": "YES",
    "EPICS_PVA_BROADCAST_PORT": "5076",
    "EPICS_PVA_DEVELOPMENT_FLAG": "0",
    "EPICS_PVA_MAINTENANCE_VERSION": "0",
    "EPICS_PVA_MAJOR_VERSION": "7",
    "EPICS_PVA_MINOR_VERSION": "1",
    "EPICS_PVA_SERVER_PORT": "5075",
    "EPICS_PVDATABASE_DEVELOPMENT_FLAG": "0",
    "EPICS_PVDATABASE_MAINTENANCE_VERSION": "1",
    "EPICS_PVDATABASE_MAJOR_VERSION": "4",
    "EPICS_PVDATABASE_MINOR_VERSION": "5",
    "EPICS_PVD_DEVELOPMENT_FLAG": "0",
    "EPICS_PVD_MAINTENANCE_VERSION": "1",
    "EPICS_PVD_MAJOR_VERSION": "8",
    "EPICS_PVD_MINOR_VERSION": "0",
    "EPICS_QSRV_ABI_MAJOR_VERSION": "1",
    "EPICS_QSRV_ABI_MINOR_VERSION": "1",
    "EPICS_QSRV_DEVELOPMENT_FLAG": "0",
    "EPICS_QSRV_MAINTENANCE_VERSION": "2",
    "EPICS_QSRV_MAJOR_VERSION": "1",
    "EPICS_QSRV_MINOR_VERSION": "2",
    "EPICS_REPO": "file:///afs/slac/g/pcds/vol2/svn/pcds",
    "EPICS_REVISION": "0",
    "EPICS_SETUP": "/reg/g/pcds/setup",
    "EPICS_SHORT_VERSION": "7.0.3.1",
    "EPICS_SITE_TOP": "/cds/group/pcds/epics",
    "EPICS_SITE_VERSION": "2.0.1",
    "EPICS_SITE_VSTRING": "-2.0.1",
    "EPICS_VERSION": "7",
    "EPICS_VERSION_NUMBER": "7.0.3.1-2.0.1",
    "EPICS_VERSION_STRING": "\"EPICS Version 7.0.3.1-2.0.1\"",

...

    "INSTALL": "perl -CSD /cds/group/pcds/epics/base/R7.0.3.1-2.0/bin/linux-x86_64/installEpics.pl -q",
    "INSTALLS_CFG": "",
    "INSTALL_ADL": "./adl",
    "INSTALL_ALH": "./alh",
    "INSTALL_ARCHIVE": "./archive",
    "INSTALL_AUTOSAVE": "./autosave",
    "INSTALL_CFG": "./cfg",
    "INSTALL_CMD": "./db",
    "INSTALL_CONFIG": "./configure",
    "INSTALL_DB": "./db",
    "INSTALL_DBD": "./dbd",
    "INSTALL_DOC": "./doc",
    "INSTALL_EDL": "./edl",
    "INSTALL_HOST_BIN": "./bin/linux-x86_64",
    "INSTALL_HOST_LIB": "./lib/linux-x86_64",
    "INSTALL_HTML": "./html",
    "INSTALL_INCLUDE": "./include",
    "INSTALL_LIBRARY": "perl -CSD /cds/group/pcds/epics/base/R7.0.3.1-2.0/bin/linux-x86_64/installEpics.pl -q",
    "INSTALL_LOCATION": ".",
    "INSTALL_LOCATION_BIN": "./bin",
    "INSTALL_LOCATION_LIB": "./lib",
    "INSTALL_PERL_MODULES": "./lib/perl",
    "INSTALL_PKGCONFIG": "./lib/pkgconfig",
    "INSTALL_PRODUCT": "perl -CSD /cds/group/pcds/epics/base/R7.0.3.1-2.0/bin/linux-x86_64/installEpics.pl -q",
    "INSTALL_PROTO": "./db",
    "INSTALL_RESTORE": "./restore",
    "INSTALL_TEMPLATES": "./templates",

...
    "RELEASE_CFG_CONFIGS": "             /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_CA_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_NTYPES_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_DATABASE_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_PVACLIENT_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_QSRV_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_PVACCESS_MODULE /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_PVDATA_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_PVDATABASE_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_PVACCESS_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_CA_MODULE /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_LIBCOM_MODULE /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_DATABASE_MODULE /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_LIBCOM_VERSION        ",
    "RELEASE_CFG_DIR_RULES": "                     ",
    "RELEASE_CFG_TOP_RULES": "                     ",
    "RELEASE_TOPS": "BASE_MODULE_VERSION EPICS_SITE_TOP EPICS_MODULES PSPKG_ROOT EPICS_BASE AUTOSAVE_MODULE_VERSION IOCADMIN_MODULE_VERSION CAPUTLOG_MODULE_VERSION IOCADMIN AUTOSAVE CAPUTLOG",
    "RELEASE_TOPS_REVERSE": "CAPUTLOG AUTOSAVE IOCADMIN CAPUTLOG_MODULE_VERSION IOCADMIN_MODULE_VERSION AUTOSAVE_MODULE_VERSION EPICS_BASE PSPKG_ROOT EPICS_MODULES EPICS_SITE_TOP BASE_MODULE_VERSION",

...


    "INSTALL": "perl -CSD /cds/group/pcds/epics/base/R7.0.3.1-2.0/bin/linux-x86_64/installEpics.pl -q",
    "INSTALLS_CFG": "",
    "INSTALL_ADL": "./adl",
    "INSTALL_ALH": "./alh",
    "INSTALL_ARCHIVE": "./archive",
    "INSTALL_AUTOSAVE": "./autosave",
    "INSTALL_CFG": "./cfg",
    "INSTALL_CMD": "./db",
    "INSTALL_CONFIG": "./configure",
    "INSTALL_DB": "./db",
    "INSTALL_DBD": "./dbd",
    "INSTALL_DOC": "./doc",
    "INSTALL_EDL": "./edl",
    "INSTALL_HOST_BIN": "./bin/linux-x86_64",
    "INSTALL_HOST_LIB": "./lib/linux-x86_64",
    "INSTALL_HTML": "./html",
    "INSTALL_INCLUDE": "./include",
    "INSTALL_LIBRARY": "perl -CSD /cds/group/pcds/epics/base/R7.0.3.1-2.0/bin/linux-x86_64/installEpics.pl -q",
    "INSTALL_LOCATION": ".",
    "INSTALL_LOCATION_BIN": "./bin",
    "INSTALL_LOCATION_LIB": "./lib",
    "INSTALL_PERL_MODULES": "./lib/perl",
    "INSTALL_PKGCONFIG": "./lib/pkgconfig",
    "INSTALL_PRODUCT": "perl -CSD /cds/group/pcds/epics/base/R7.0.3.1-2.0/bin/linux-x86_64/installEpics.pl -q",
    "INSTALL_PROTO": "./db",
    "INSTALL_RESTORE": "./restore",
    "INSTALL_TEMPLATES": "./templates",

...

    "MAKEFILE_LIST": " /tmp/Gm9iER6U configure/CONFIG configure/RELEASE RELEASE_SITE /cds/group/pcds/epics/base/R7.0.3.1-2.0/configure/CONFIG /cds/group/pcds/epics/base/R7.0.3.1-2.0/configure/RELEASE /cds/group/pcds/epics/base/R7.0.3.1-2.0/configure/CONFIG_COMMON /cds/group/pcds/epics/base/R7.0.3.1-2.0/configure/CONFIG_FILE_TYPE /cds/group/pcds/epics/base/R7.0.3.1-2.0/configure/CONFIG_BASE /cds/group/pcds/epics/base/R7.0.3.1-2.0/configure/CONFIG_SITE /cds/group/pcds/epics/base/R7.0.3.1-2.0/configure/CONFIG_BASE_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/configure/os/CONFIG.linux-x86_64.Common /cds/group/pcds/epics/base/R7.0.3.1-2.0/configure/os/CONFIG.UnixCommon.Common /cds/group/pcds/epics/base/R7.0.3.1-2.0/configure/os/CONFIG_SITE.linux-x86_64.Common /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_PVDATA_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_CA_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_LIBCOM_MODULE /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_PVDATABASE_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_DATABASE_MODULE /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_NTYPES_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_PVACCESS_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_CA_MODULE /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_LIBCOM_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_DATABASE_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_PVACLIENT_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_QSRV_VERSION /cds/group/pcds/epics/base/R7.0.3.1-2.0/cfg/CONFIG_PVACCESS_MODULE configure/CONFIG_SITE configure/RULES_TOP /cds/group/pcds/epics/base/R7.0.3.1-2.0/configure/RULES_TOP /cds/group/pcds/epics/base/R7.0.3.1-2.0/configure/RULES_DIRS /cds/group/pcds/epics/base/R7.0.3.1-2.0/configure/RULES_COMMON",


    "UNINSTALL_DIRS": "./archive ./autosave ./cfg ./db ./dbd ./doc ./html ./include  ./templates ./adl ./alh ./cfg ./edl ./lib/perl ./lib/pkgconfig ./restore ./db ./db",

```

</details>

## Motivation and Context
* Introspecting parts of the EPICS build system
* Will be used in other tools for auto-downloading/building dependencies
* Will meet up with https://github.com/pcdshub/ioc_machine_core/pull/4 at some point
 
## How Has This Been Tested?
* Included test suite

## Where Has This Been Documented?
* This PR text

### CODE REVIEW

**Comment by klauer** [comment]
This still needs:
* Some tools for picking out dependencies (existing or not)
* A fix for travis (make version difference? code bug? let's see...)
* Maybe an additional command-line tool for recursively finding dependencies and building the tree

**Comment by klauer** [comment]
Extracting IOC dependency information for display works:
image

Support for building and downloading may be in an external tool, I think...

**Comment by klauer** [comment]
I think this is good enough as a first pass. Testing is lacking on the `DependencyGroup` stuff but I'd like to focus on that down the line when I get around to using this a bit more.

➡️  Requesting at least a smell check, leaving level of review up to reviewer's whim

**Comment by klauer** [comment]
OK, pending tests passing, this is really ready this time.
And perhaps this better represents the _insanity that is EPICS module dependencies_ than the above simple example:
> ads-ioc R0.5.0:
image

**Comment by ZLLentz** [comment]
Wow that's a gnarly dependency tree

File: whatrecord/bin/deps.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+"""
+"whatrecord info" is used to get PV information from the whatrecord server.
```
[the remark this answers is not in the document]
Comment (klauer) [reply whatrecord/bin/deps.py]: Thanks! Fixed up in [022165a](https://github.com/pcdshub/whatrecord/pull/124/commits/022165aecff1e21087a5a0f05fa642201007cc7a)

**Review by tangkong — APPROVED** [review]
I'm not experienced enough with EPICS to be a great judge of this functionality, but it looks useful and well put-together to me.  Tests look good too.

I should really play around with this more to get a better grasp on things, but at this moment it LGTM

File: whatrecord/makefile.py [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+    #: Modules (etc.) which depend on this instance
+    dependents: List[pathlib.Path] = field(default_factory=list)
+    #: Modules (etc.) are required for this instance
+    dependencies: List[pathlib.Path] = field(default_factory=list)
+    #: The "root" node.  ``None`` to refer to itself.
+    root: Optional[DependencyGroup] = field(default=None, metadata=apischema.metadata.skip)
+
+    @classmethod
+    def from_makefile(
+        cls,
```
Comment (ZLLentz) [inline whatrecord/makefile.py:445]: nitpick: this is a `@classmethod` that doesn't use the `cls` variable

**Review by ZLLentz — APPROVED** [review]
I didn't fully grasp how this works at the core (lack of Makefile knowledge), but the result is good and I only had minor nitpicks

File: whatrecord/makefile.py (hunk 2 above)
Comment (klauer) [reply whatrecord/makefile.py:445]: Not a nitpick - a bug.

**Pull request merged**